apps/passport-server/src/services/devconnectPretixSyncService.ts

```diff
@@ -46,126 +48,259 @@ interface EventData {
   tickets: DevconnectPretixOrder[];
 }
 
+type SyncPhase = "fetching" | "validating" | "saving";
+
+type FetchOutcome = "complete" | "rate-limited";
+
+interface SyncErrorCause {
+  success: false;
+  phase: SyncPhase;
+  error: Error;
+  organizerId: string;
+}
+
+function errorCause(
+  phase: SyncPhase,
+  organizerId: string,
+  originalError: any
+): SyncErrorCause {
+  if (originalError instanceof Error) {
+    return { success: false, phase, error: originalError, organizerId };
+  } else {
+    throw new Error(`originalError is not an error`);
+  }
+}
+
 /**
- * Responsible for syncing users from Pretix into an internal representation.
+ * We have a concept of a sync "run", which occurs on an interval.
+ * A run can fail, can complete synchronization, or can be suspended due
+ * to rate-limiting on requests to Pretix.
+ *
+ * In order to do a full synchronization for a given organizer, multiple
+ * runs may be required if rate-limiting makes it impossible to fetch
+ * all of the necessary data in a single run.
  */
-export class DevconnectPretixSyncService {
-  private static readonly SYNC_INTERVAL_MS = 1000 * 60;
+interface SyncRunSuccess {
+  success: true;
+  organizerId: string;
+  outcome: "complete" | "rate-limited";
+}
+
+type SyncResult = SyncRunSuccess | SyncErrorCause;
+
+/**
+ * This class implements a rate-limited fetcher. It provides a function
+ * which can be passed to the Pretix fetch functions, in place of the
+ * native 'fetch' API call. Behind the scenes, this function will
+ * increment a counter. When the counter exceeds the limit, the call to
+ * the fetcher function will suspend until the counter is reset. When
+ * this happens, an event is raised. In this system, we use this event
+ * to mark a sync run as "rate-limited", allowing us to complete the run
+ * and resume fetching the remaining data on the next run, when we will
+ * reset the counter.
+ */
+class LimitedFetcher extends EventEmitter {
```

Comment:
in general I avoid using `EventEmitter` this way because 1) it's not strongly typed and 2) it's not evident without reading the code what all the possible events this emitter could emit are.

i recommend avoiding this